### PR TITLE
[NVIDIA] Support GQA and MQA for `jax.nn.dot_product_attention`.

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables the `jax.nn.dot_product_attention` to support GQA and MQA. In particular, it allows the Key/Value tensors to have different number of heads from the Query. So:

* If KV and Q have same number of heads => normal attention
* If KV has no head or 1 head => MQA
* If Q has a multiple of number of heads of KV => GQA

### CODE REVIEW

**Comment by kaixih** [comment]
@sbodenstein @superbobry to review.

**Comment by superbobry** [comment]
@kaixih how urgent is this? I would rather have Sebastian review first, but he is out at the moment.

**Review by zhangqiaorjc — APPROVED** [review]

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -844,24 +871,29 @@ def dot_product_attention(
...
   Args:
     query: query array; shape :code:`(BTNH)`
-    key: key array; shape :code:`(BSNH)`
-    value: value array; shape :code:`(BSNH)`
-    bias: optional, bias array to be added to logits; shape broadcastable to
-      :code:`(BNTS)`.
+    key: key array: shape :code:`(BSKH)`. When `K` equals `N`, multi-headed
+      attention (MHA) is performed. Otherwise, when `N` is a multiple of `K`,
+      grouped query attention (GQA) is performed. Additionally, it supports the
+      shape :code:`(BSH)` for multi-query attention (MQA).
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:889]: Why do we need to support a different rank here? Isn't MQA just K=1? (ie group size 1) Complicates a lot of code having to check for this, and will make jaxtyping type signatures more complicated if we ever want to support them.

Comment (kaixih) [reply jax/_src/nn/functions.py]: Sure. Removed.

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -844,24 +871,29 @@ def dot_product_attention(
   :code:`probs` as the output of :math:`softmax`.
 
   Throughout this function, we utilize the following uppercase letters to
-  represent the shape of array:
+  represent the shape of array::
```
[the remark this answers is not in the document]
Comment (kaixih) [reply jax/_src/nn/functions.py:874]: I notice that in the generated doc the items in the legend are reformated to one line ([here](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.dot_product_attention.html#jax.nn.dot_product_attention)). Then, I found it seems we can use the double colons to preserve the format in the code-like block. For example: code [here](https://github.com/google/jax/blob/0d5dae09ffe30ab72c6600b3650a41f8915aa2b4/jax/random.py#L37) will generate doc like [here](https://jax.readthedocs.io/en/latest/jax.random.html#prng-keys).

[the remark this answers is not in the document]
Comment (sbodenstein) [reply jax/_src/nn/functions.py:874]: OK thanks!

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -844,24 +871,29 @@ def dot_product_attention(
   :code:`probs` as the output of :math:`softmax`.
```
Comment (sbodenstein) [inline jax/_src/nn/functions.py:872]: Would it make sense to give the arxiv references to MHA, GQA and MQA? We use these terms, but its useful for users to know what they mean precisely.

File: tests/nn_test.py [hunk 4]
```diff
@@ -68,8 +70,8 @@ def testDotProductAttentionInfer(self, dtype, use_bias, causal_mode, impl):
     B, S, T, N, H = 2, 128, 128, 4, 32
     keys = random.split(random.PRNGKey(0), 4)
     Q = random.normal(keys[0], (B, T, N, H), dtype)
-    K = random.normal(keys[1], (B, S, N, H), dtype)
-    V = random.normal(keys[2], (B, S, N, H), dtype)
+    K = random.normal(keys[1], (B, S, N // group_num, H), dtype)
```
Comment (sbodenstein) [inline tests/nn_test.py:73]: Can we add a test for GQA/MQA correctness where we just replicate the head_dim of key/value and use standard MHA? This also makes it clear exactly what this is doing.

**Review by sbodenstein — COMMENTED** [review]
I think the XLA implementation is not the most efficient, but we can improve this in a followup CL.

File: tests/nn_test.py (hunk 4 above)
Comment (kaixih) [reply tests/nn_test.py]: Done. PTAL.

File: jax/_src/nn/functions.py (hunk 3 above)
Comment (kaixih) [reply jax/_src/nn/functions.py:872]: Sure. Added the paper urls.

**Review by sbodenstein — COMMENTED** [review]
LGTM

**Review by superbobry — APPROVED** [review]

**Comment by superbobry** [comment]
Please squash the commits :)

**Comment by kaixih** [comment]
> Please squash the commits :)

Done. PTAL.

**Review by zhangqiaorjc — APPROVED** [review]

**Pull request merged**